File: elixir/baffling-birthdays/lib/baffling_birthdays.ex
defmodule BafflingBirthdays do
  @moduledoc """
  Estimate the probability of shared birthdays in a group of people.
  """
  @sims 10000
  def shared_birthday?(ls), do: Enum.map(ls, &extract_md/1) |> Enum.group_by(& &1) |> Enum.any?(&shared_group?/1)
  def random_birthdates(sz), do: Enum.take(Stream.repeatedly(&rand_date/0), sz)
  def estimated_probability_of_shared_birthday(sz), do: Enum.count(1..@sims,fn _ -> run_sim(sz) end) / @sims * 100.0

  defp rand_date(), do: Date.add(~D[2025-01-01], :rand.uniform(365) - 1)
  defp extract_md(%{month: m, day: d} = _date), do: {m, d}
  defp shared_group?({_key, group}), do: length(group) > 1
  defp run_sim(sz), do: shared_birthday?(random_birthdates(sz))
end

File: csharp/baffling-birthdays/BafflingBirthdays.cs
public static class BafflingBirthdays
{
    private static DateOnly RandomDateIn(int year)
    {
        var month = Random.Shared.Next(1, 13);
        var day = Random.Shared.Next(1, DateTime.DaysInMonth(year, month) + 1);
        return new DateOnly(year, month, day);
    }
        
    public static DateOnly[] RandomBirthdates(int num) => // just use 2025 for the year
        Enumerable.Range(0, num).Select(_ => RandomDateIn(2025)).ToArray();        
    
    public static bool SharedBirthday(DateOnly[] birthdays)
    {
        var seen = new HashSet<(int, int)>();
        foreach (var date in birthdays)
            if (!seen.Add((date.Month, date.Day))) 
                return true;
        return false;
    }
    
    public static double EstimatedProbabilityOfSharedBirthday(int numberOfBirthdays) =>
        (Enumerable.Range(0, 10000).Count(_ => SharedBirthday(RandomBirthdates(numberOfBirthdays))) / 
        (double)10000 ) * 100.0;
}
